File: AA/Lab2/monotonia.py
n = int(input())
point = [int(i) for i in input().split()]

points = [point]
lefter = point
right = point
top = point
bottom = point
l = 0
r = 0
t = 0
b = 0

for i in range(1,n):
    point = [int(i) for i in input().split()]
    points.append(point)

    if point[0] < lefter[0]:
        lefter = point
        l = i

    if point[0] > right[0]:
        right = point
        r = i

    if point[1] < bottom[1]:
        bottom = point
        b = i

    if point[1] > top[1]:
        top = point
        t = i

points.append(points[0])

xMonoton = True
yMonoton = True

i = b + 1
while i % n != t:
    if points[i % n][1] < points[i % n - 1][1]:
        yMonoton = False
        break

    i += 1

if yMonoton:
    i = b - 1
    while i != t:
        if i == -1:
            i = n - 1
        if i == t:
            break
        
        if points[i][1] < points[i + 1][1]:
            yMonoton = False
            break

        i -= 1

i = l + 1
while i % n != r:
    if points[i % n][0] < points[i % n - 1][0]:
        xMonoton = False
        break

    i += 1

if xMonoton:
    i = l - 1
    while i != r:
        if i == -1:
            i = n - 1
        if i == r:
            break
        
        if points[i][0] < points[i + 1][0]:
            xMonoton = False
            break

        i -= 1

if xMonoton:
    print("YES")
else:
    print("NO")

if yMonoton:
    print("YES")
else:
    print("NO")

File: AA/Lab2/monotonia.cpp
#include <bits/stdc++.h> 
using namespace std;
typedef pair< int,  int> pii;


vector<pii> points;
pii lefter, righter, top, bottom;
int l, r, t, b;

int main() {
    int n;
     int x, y;

    cin >> n;
    cin >> x >> y;
    lefter = make_pair(x, y);
    righter = lefter;
    top = lefter;
    bottom = lefter;

    points.push_back(lefter);

    for(int i = 1; i < n; i++) {
        cin >> x >> y;
        pii point = make_pair(x, y);

        if(point.first < lefter.first) {
            lefter = point;
            l = i;
        } 
         if(point.first > righter.first) {
            righter = point;
            r = i;
        } 
         if(point.second < bottom.second) {
            bottom = point;
            b = i;
        } 
         if(point.second > top.second) {
            top = point;
            t = i;
        }

        points.push_back(point);
    }
    points.push_back(points[0]);

    // cout << l << " " << t << " " << r << " " << b << '\n';
    // cout << "\n----------------------------\n\nX Monotonie 1:\n";

    bool xMonoton = true, yMonoton = true;

    //y monotonie

// cout << "\n----------------------------\n\nY Monotonie 1:\n";
    for(int i = b + 1; i % n != t; i++) {
        // cout << points[i].first << " " << points[i].second << '\n';
        if(points[i % n].second < points[i % n - 1].second) {
            // cout << "here\n";
            yMonoton = false;
            break;
        }
    }
    
// cout << "\n----------------------------\n\nY Monotonie 2:\n";
    if(yMonoton) 
        for(int i = b - 1; i != t; i--) 
            {
                if(i == -1) i = n - 1;
                if(i == t) break;
                // cout << points[i].first << " " << points[i].second << '\n';
                if(points[i].second < points[i + 1].second) {
                    // cout << "here\n";
                    yMonoton = false;
                    break;
                }
            }

    //x monotonie
    for(int i = l + 1; i % n != r; i++) {
        // cout << points[i].first << " " << points[i].second << '\n';
        if(points[i % n].first < points[(i % n) - 1].first) {
            // cout << "here\n";
            xMonoton = false;
            break;
        }
    }
        
    // cout << "\n----------------------------\n\nX Monotonie 2:\n";
    if(xMonoton) 
        for(int i = l - 1; i != r; i--) 
            {
                if(i == -1) i = n - 1;
                if(i == r) break;
                // cout << points[i].first << " " << points[i].second << '\n';
                if(points[i].first < points[i + 1].first) {
                    // cout << "here\n";
                    xMonoton = false;
                    break;
                }
            }
    
    


    cout << (xMonoton ? "YES\n" : "NO\n");
    cout << (yMonoton ? "YES\n" : "NO\n");

    return 0;
}